### PR TITLE
fix(mcp): reject unsafe request-scoped header values

### LINKED ISSUE #5065: [bug] An MCP credential with a trailing newline gets echoed into a model-visible tool error
### Before you start

- [x] I searched [existing issues](https://github.com/bytedance/deer-flow/issues?q=is%3Aissue) and this is not a duplicate.
- [ ] I can reproduce this on the latest `main`.

### Problem summary

If a secret passed in config.context.secrets (or a user_auth credential) contains a character that is illegal in an HTTP header value, such as a trailing newline, the HTTP client rejects it with an exception that echoes the full value, and that message becomes a model-visible tool error. The credential ends up in the prompt, the checkpoint, and traces.

### Affected area(s)

MCP

### What happened?

mcpServers.<server>.headers_from_context injects values from the run request's config.context.secrets into HTTP headers on every tool call (docs/MCP_SERVER.md promises those values are "stripped from persisted run configuration, API responses, and trace payloads"). The interceptor validates that a mapped key is present and non-empty, but not that the value can actually travel as an HTTP header value.

A trailing newline is the usual way this happens in practice: a caller reads a token from a file or a Kubernetes secret mount and forgets to strip it. The interceptor passes the value straight to httpx, and h11 rejects it at send time with:

    httpx.LocalProtocolError: Illegal header value b'Bearer sk-tenant-secret-abc123\n'

Note the full credential in the message. ToolErrorHandlingMiddleware then converts the exception into a model-visible ToolMessage ("Error: Tool 'x' failed with LocalProtocolError: Illegal header value b'Bearer sk-tenant-secret-abc123\n'. ..."), so the secret ends up in the model context, in the thread checkpoint, and in trace payloads. These are exactly the places the feature promises the value never goes (see the module docstring of packages/harness/deerflow/mcp/context_headers.py).

user_auth credentials have the same problem: a $ENV_VAR value picked up from a CRLF env-file, or set from a token file with a trailing newline, produces the same exception with the credential echoed back.

Values that httpx accepts but h11 refuses include CR, LF, NUL, and leading/trailing whitespace; values outside Latin-1 fail earlier inside httpx with a UnicodeEncodeError. In every case the tool call was going to fail anyway. The bug is that the failure message carries the secret into surfaces that outlive the request.

### Expected behavior

A credential that cannot be sent as an HTTP header value should fail closed inside the interceptor, with an error that names the offending secret key (so the caller knows what to fix) but never repeats the value. This matches how the interceptors already treat a missing or empty credential.

### Steps to reproduce

```
# backend/, after uv sync. Minimal demonstration of the echo:
import asyncio, httpx

async def main():
    async with httpx.AsyncClient() as client:
        # what the MCP HTTP transport does with the injected header
        await client.post("http://127.0.0.1:9/mcp",
                          headers={"authorization": "Bearer sk-tenant-secret-abc123\n"})

asyncio.run(main())

# httpx.LocalProtocolError: Illegal header value b'Bearer sk-tenant-secret-abc123\n'
# End to end: configure an `http` MCP server with
#   headers_from_context: {"headers": {"Authorization": "tenant_token"}}
# start a run whose request carries
#   config.context.secrets: {"tenant_token": "Bearer sk-tenant-secret-abc123\n"}
# and let the agent call any tool from that server. The resulting tool error
# shown to the model (and persisted with the thread) contains the full secret.
# The interceptor never rejects the value: _request_secrets ->
# resolved[header_name] = value in packages/harness/deerflow/mcp/context_headers.py
# only checks for presence, and ToolErrorHandlingMiddleware._build_error_message
# copies str(exc) into the ToolMessage.
```

### Relevant logs

```shell
httpx.LocalProtocolError: Illegal header value b'Bearer sk-tenant-secret-abc123\n'
  (raised from h11, via langchain-mcp-adapters' streamable HTTP session)
ToolMessage(status="error"): "Error: Tool 'shared-api_act' failed with LocalProtocolError: Illegal header value b'Bearer sk-tenant-secret-abc123\n'. Continue with available context, or choose an alternative tool."
```

### How are you running DeerFlow?

Local (make dev)

### Operating system

Linux

### Platform details

x86_64, bash

### Python version

Python 3.12.3

### Node.js version

v24.18.0

### pnpm version

_No response_

### uv version

0.11.32

### Git state

_No response_

### Support bundle summary

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by gkmraju**
Fixes #5065

## Why

Request-scoped MCP secrets can contain malformed header values. If those values reach the HTTP transport, downstream exceptions may echo the malformed credential text. This can expose secret material in logs or error messages.

## What changed

- Reject request-scoped MCP header values containing control characters or newlines.
- Reject surrounding whitespace and values that cannot be represented safely as Latin-1 HTTP header content.
- Raise validation errors that identify only the configured secret key name, never the secret value.
- Add regression coverage for unsafe values.

## Surface area

- [x] **Backend API**
- [ ] **Docs / tests / CI only**  

## Bug fix verification

Test path: the new MCP header-validation regression test in this PR.

Executable tests were not run locally through the connected integration, so I am not claiming local green status. Upstream CI should validate the branch.

## Validation

Source, issue context, duplicate PRs, and the focused diff were inspected. No executable tests or lint were run locally.

## AI assistance

**Tool(s) used:** OpenAI GPT-5.6 Sol

**How you used it:** Assisted with issue analysis, implementation, regression-test design, and diff review.

### CODE REVIEW

File: backend/packages/harness/deerflow/mcp/context_headers.py [hunk 1]
```diff
@@ -77,6 +77,25 @@ def _request_secrets(request: Any) -> dict[str, str]:
     return extract_request_secrets(getattr(runtime, "context", None))
 
 
+def _is_safe_header_value(value: str) -> bool:
```
Comment (willem-bd) [inline backend/packages/harness/deerflow/mcp/context_headers.py:80]: This fixes only the `headers_from_context` path, but #5065 also names `user_auth`, and that path is still reachable.

`user_scoped_auth.py:119-124` injects `credential` - the `$ENV_VAR`-resolved value from `user_auth.users`, which is exactly the CRLF-env-file / trailing-newline source the issue describes - into `user_auth.header` via `apply_header_overrides`, with no validation. `apply_header_overrides` (`mcp/headers.py`) does no value checks, so a trailing newline still reaches httpx and `LocalProtocolError` still echoes the credential into a model-visible ToolMessage.

Because the PR says `Fixes #5065`, the issue will auto-close with that leak still live end-to-end. Either run the `user_scoped_auth` credential through `_is_safe_header_value` as well, or drop the `Fixes` keyword and note the remaining surface in the description.

**Review by willem-bd — COMMENTED** [review]
Reviewed at cfbb20b6. The core goal is met: I confirmed against the pinned stack (httpx 0.28.1 -> h11 0.16.0) that the values h11 actually rejects - interior CR/LF/vertical-tab, NUL, and leading/trailing whitespace - are exactly the ones `_is_safe_header_value` catches, and the new `ToolException` names only secret *keys*, never values. `tests/test_mcp_context_header_validation.py` passes (55/55 together with the existing context-headers suite).

Four things worth addressing, in rough priority order: the `user_auth` path from #5065 is still unpatched, so `Fixes` will close the issue with that leak live; the docstring's Latin-1 rationale is factually wrong for the pinned transport, making the check stricter than it needs to be in ways the test suite mislabels; the new tests duplicate an existing suite rather than extending it; and the unconditional deny silently narrows the documented `on_missing: passthrough` opt-out.

File: backend/packages/harness/deerflow/mcp/context_headers.py [hunk 2]
```diff
@@ -77,6 +77,25 @@ def _request_secrets(request: Any) -> dict[str, str]:
...
+    """Return whether a request-scoped secret is safe to pass as an HTTP header.
+
+    Reject values that the HTTP transport may reject while echoing the original
+    bytes in its exception. Credential values should not contain control
+    characters or surrounding whitespace, and httpx encodes header strings as
+    Latin-1 before handing them to the HTTP protocol implementation.
```
Comment (willem-bd) [inline backend/packages/harness/deerflow/mcp/context_headers.py:86]: This rationale does not hold for the pinned transport, which changes what the function is really doing. Against the lockfile versions (httpx 0.28.1, h11 0.16.0):

- `httpx.Headers.__setitem__` (`httpx/_models.py:309`) encodes header values as **UTF-8**, not Latin-1, so a non-Latin-1 value never raises `UnicodeEncodeError`.
- h11 validates with `_field_value_re` = `[not-NUL-or-ws]+(?:[ \t]+[not-NUL-or-ws]+)*` under `fullmatch` (`h11/_headers.py:74,166`), which **accepts** interior tabs, DEL, other C0 controls such as `\x01`, C1 controls such as `\x85`, and UTF-8-encoded non-ASCII.

I verified this end-to-end against the venv: `'Bearer \u20ac'`, `'Bearer\ta b'`, `'Bearer a\x7fb'`, `'Bearer\x01a b'` and `'Bearer\x85ab'` all reach the wire today, but `_is_safe_header_value` now rejects them - so a credential that was working starts failing hard.

Being stricter than the transport is a perfectly defensible fail-closed posture for a credential, but the docstring should say that rather than asserting a transport constraint that does not exist; the next reader will trust it. `test_unsafe_header_values_are_rejected` pins the wrong model too - `'Bearer secret-\u20ac'` is not transport-unsafe, so it belongs under a "deliberately stricter than the transport" framing rather than being presented as a value the transport rejects.

File: backend/packages/harness/deerflow/mcp/context_headers.py [hunk 3]
```diff
@@ -128,15 +147,33 @@ async def context_headers_interceptor(request: Any, handler: Any) -> Any:
...
             value = secrets.get(secret_key, "")
-            if value:
+            if value and _is_safe_header_value(value):
                 resolved[header_name] = value
+            elif value:
+                invalid.append(secret_key)
             else:
                 missing.append(secret_key)
 
+        if invalid:
```
Comment (willem-bd) [inline backend/packages/harness/deerflow/mcp/context_headers.py:162]: This deny fires regardless of `on_missing`. The documented contract on `McpContextHeadersConfig.on_missing` is that `passthrough` "forwards the request with the server's static headers" when a mapped value is absent or empty, so an operator who opted out of fail-closed now gets a hard deny for values that are present-but-malformed - new behaviour that is neither documented nor tested.

I think denying is the right call (falling back to the server's static credential would silently switch the caller's identity), but please pin it with a test next to `test_on_missing_passthrough_keeps_static_headers` and mention it in the `on_missing` field description, so `passthrough` reads as "passthrough for absent/empty, always deny for malformed".

File: backend/tests/test_mcp_context_header_validation.py [hunk 4]
```diff
@@ -0,0 +1,67 @@
+"""Regression tests for request-scoped MCP credential header validation."""
```
Comment (willem-bd) [inline backend/tests/test_mcp_context_header_validation.py:1]: `backend/tests/test_mcp_context_headers.py` is already the test home for this module - it defines `_config()`, `_request()` and `_echo_handler()` plus the sibling deny/leak cases (`test_empty_secret_value_is_denied`, `test_deny_message_does_not_leak_other_secret_values`, `test_on_missing_passthrough_keeps_static_headers`). This file re-declares near-identical helpers, so the interceptor's deny behaviour is now split across two files and they will drift; this one's `_config()` already omits the static `headers` the other one sets.

Folding these cases into the existing file keeps the helpers shared and puts the new assertions next to the deny tests they extend.

**Comment by WillemJiang** [comment]
@gkmraju  please click the CLA assistant button to sign the CLA first.
